pkg/volume/csi/csi_plugin.go

```diff
@@ -901,33 +895,25 @@ func highestSupportedVersion(versions []string) (*utilversion.Version, error) {
 			theErr = err
 			continue
 		}
-		if currentHighestVer.Major() > 1 {
-			// CSI currently only has version 0.x and 1.x (see https://github.com/container-storage-interface/spec/releases).
-			// Therefore any driver claiming version 2.x+ is ignored as an unsupported versions.
-			// Future 1.x versions of CSI are supposed to be backwards compatible so this version of Kubernetes will work with any 1.x driver
-			// (or 0.x), but it may not work with 2.x drivers (because 2.x does not have to be backwards compatible with 1.x).
-			continue
-		}
 		if highestSupportedVersion == nil || highestSupportedVersion.LessThan(currentHighestVer) {
 			highestSupportedVersion = currentHighestVer
 		}
 	}
 
-	if highestSupportedVersion != nil {
-		return highestSupportedVersion, nil
+	if highestSupportedVersion == nil {
+		return nil, fmt.Errorf("could not find a highest supported version from versions (%v) reported by this driver: %v", versions, theErr)
 	}
-	return nil, fmt.Errorf("None of the CSI versions (%v) reported by this driver are supported : %v", versions, theErr)
-}
 
-// Only drivers that implement CSI 0.x are allowed to use deprecated socket dir.
-func isDeprecatedSocketDirAllowed(versions []string) bool {
-	for _, version := range versions {
-		if isV0Version(version) {
-			return true
-		}
+	if highestSupportedVersion.Major() != 1 {
```

Comment:
What happens if a driver supports 2.x and 1.x? In the future, if we have a 2.x, the driver Daemonset could theoretically run with a driver that supports both 1.x and 2.x.

Reply:
I think we would have to go in and update the validation code. I think the theory is that v2.x will have breaking changes and we wouldn't want to just "quietly accept" that version in older versions of k8s. If we do decide that older versions of k8s support `v2.x+` in the future we can always cherry-pick a relaxation in validation.

Reply:
if the driver supports 1.x and 2.x, current versions of kube should allow it and choose to use 1.x, right? does this code do that?

Reply:
oh just understood this - yeah fixing. If v2.0 AND v1.0 are supported we should return the highest v1.0 version. Fixing now

Reply:
yep, fixed. In the code above it will skip any >v1 versions as a `highestSupportedVersion` candidate - so if a driver supports v2.0 and v1.5 the `highestSupportedVersion` will be `v1.5`